src/coreclr/jit/codegenloongarch64.cpp

```diff
@@ -0,0 +1,9260 @@
+// genStoreLclTypeSIMD12: store a TYP_SIMD12 (i.e. Vector3) type field.
+// Since Vector3 is not a hardware supported write size, it is performed
+// as two stores: 8 byte followed by 4-byte.
+//
+// Arguments:
+//    treeNode - tree node that is attempting to store TYP_SIMD12 field
+//
+// Return Value:
+//    None.
+//
+void CodeGen::genStoreLclTypeSIMD12(GenTree* treeNode)
+{
+    assert(!"unimplemented on LOONGARCH yet");
+}
+
+#endif // FEATURE_SIMD
+
+/*****************************************************************************
+ * Unit testing of the LOONGARCH64 emitter: generate a bunch of instructions into the prolog
+ * (it's as good a place as any), then use COMPlus_JitLateDisasm=* to see if the late
+ * disassembler thinks the instructions as the same as we do.
+ */
+
+// Uncomment "#define ALL_LOONGARCH64_EMITTER_UNIT_TESTS" to run all the unit tests here.
+// After adding a unit test, and verifying it works, put it under this #ifdef, so we don't see it run every time.
+//#define ALL_LOONGARCH64_EMITTER_UNIT_TESTS
+
+#if defined(DEBUG)
+void CodeGen::genLoongArch64EmitterUnitTests()
+{
+    if (!verbose)
+    {
+        return;
+    }
+
+    if (!compiler->opts.altJit)
+    {
+        // No point doing this in a "real" JIT.
+        return;
+    }
+
+    // Mark the "fake" instructions in the output.
+    printf("*************** In genLoongArch64EmitterUnitTests()\n");
+
+    printf("*************** End of genLoongArch64EmitterUnitTests()\n");
+}
+#endif // defined(DEBUG)
+
+//------------------------------------------------------------------------
+// genStackPointerConstantAdjustment: add a specified constant value to the stack pointer.
+// No probe is done.
+//
+// Arguments:
+//    spDelta                 - the value to add to SP. Must be negative or zero.
+//    regTmp                  - an available temporary register that is used if 'spDelta' cannot be encoded by
+//                              'sub sp, sp, #spDelta' instruction.
+//                              Can be REG_NA if the caller knows for certain that 'spDelta' fits into the immediate
+//                              value range.
+//
+// Return Value:
+//    None.
+//
+void CodeGen::genStackPointerConstantAdjustment(ssize_t spDelta, regNumber regTmp)
+{
+    assert(spDelta < 0);
+
+    // We assert that the SP change is less than one page. If it's greater, you should have called a
+    // function that does a probe, which will in turn call this function.
+    assert((target_size_t)(-spDelta) <= compiler->eeGetPageSize());
+
+    if (-2048 <= spDelta && spDelta < 0)
+        GetEmitter()->emitIns_R_R_I(INS_addi_d, EA_PTRSIZE, REG_SPBASE, REG_SPBASE, spDelta);
+    else
+    {
+        GetEmitter()->emitIns_R_R_I(INS_srai_d, EA_PTRSIZE, REG_SPBASE, REG_SPBASE, 3);
+        GetEmitter()->emitIns_R_R_I(INS_addi_d, EA_PTRSIZE, REG_SPBASE, REG_SPBASE, spDelta >> 3);
+        GetEmitter()->emitIns_R_R_I(INS_slli_d, EA_PTRSIZE, REG_SPBASE, REG_SPBASE, 3);
```

Comment:
As I mentioned elsewhere, I think this adjustment of SP is either illegal or a very bad idea, as it leaves SP in a corrupt state after the srai.d/addi.d instructions.

Reply:
Thanks very much.